File: ImageCompressor.m
clear all
clc

ImageRead = imread('1.png');
X = zeros((size(ImageRead,1)*size(ImageRead,2)/64), 64);
k=0;
for i=1:8:size(ImageRead,1)
    for j=1:8:size(ImageRead,1)
       one_block_values= ImageRead(i:i+8-1,j:j+8-1);
       k=k+1;
       X(k,:)=reshape(one_block_values',1,64);
    end
end
mean_x=mean(X);
for i=1:64
	X(:,i)=X(:,i)-mean_x(i);
end
co=zeros(64);
for i=1:64
    for j=1:64 
       co(i,j)=sum(X(:,i).*X(:,j))/(size(ImageRead,1)*size(ImageRead,2)/64);
    end
end
[V,D]=eig(co);
for s=[5 10 20 30 40 50 60]
    flippedV=fliplr(V);
    standardizedX=flippedV'/norm(flippedV')*X';
    flippedV(:,64-s+1:end)=0;
    retrivedX=(flippedV/norm(flippedV))*standardizedX;
    retrivedX=retrivedX';
    for i=1:64
		retrivedX(:,i)=retrivedX(:,i)+mean_x(i);
    end
    FinalImage=zeros(size(ImageRead,1),size(ImageRead,2));
    k=1;
    for i=1:8:size(ImageRead,1)
        for j=1:8:size(ImageRead,1)
            FinalImage(i:i+8-1,j:j+8-1)=reshape(retrivedX(k,:),8,8)';
            k=k+1;
        end
    end
    figure
    imshow(FinalImage,[]);
	print(['converted/',num2str(s)],'-dpng')
end

File: ImageCompressor.py
from PIL import Image
import numpy as np
import sklearn.decomposition as deco
import pylab
import threading

def matrixMul(X, Y):
	result = np.zeros((len(X), len(Y[0])))
	t1 = threading.Thread(target=calculator, args = (X,Y,result,0))
	t1.start()
	t2 = threading.Thread(target=calculator, args = (X,Y,result,16))
	t2.start()
	t3 = threading.Thread(target=calculator, args = (X,Y,result,32))
	t3.start()
	t4 = threading.Thread(target=calculator, args = (X,Y,result,48))
	t4.start()
	t1.join()
	t2.join()
	t3.join()
	t4.join()
	print(result)
	return result
	
def calculator(X, Y, result, i):
	for m in range(i, i+16):
		for j in range(len(Y[0])):
			for k in range(len(Y)):
			   result[m][j] += X[m][k] * Y[k][j]
		print(m)

im_original = np.asarray(Image.open('1.png').convert('L'))				#converting image into N*M matrix

linear_matrix=[]

height = len(im_original)
width = len(im_original[0])
for i in range(0, height/8):
	for j in range(0, width/8):
		index_i = i*8
		index_j = j*8
		linear_matrix.append(im_original[index_i:index_i+8, index_j:index_j+8].reshape(1, 64))			#converting image matrix into linear_matrix
x = []
for i in range(0, len(linear_matrix)):
	x.append(linear_matrix[i][0])

std_x = np.std(x, 0)
mean_x = np.mean(x, axis=0)
x = (x - np.mean(x, axis=0)) / np.std(x, 0) # You need to normalize your data first
cov = np.ma.cov(x, rowvar=False)

D, V = np.linalg.eig(cov)

for k in (5, 10, 20, 30, 40, 50, 60):
	tV = np.transpose(V)
	x_new = matrixMul((tV / np.linalg.norm(tV)), np.transpose(x))
	for c in range(0, k):
		for r in range(0, 64):
			V[r][c] = 0
	x_Ret = matrixMul((V / np.linalg.norm(V)), x_new)
	x_Ret = np.transpose(x_Ret)
	x_Ret = (x_Ret * std_x) + mean_x
	x_Ret = x_Ret.reshape(512, 512)
	pylab.figure()
	pylab.gray()
	pylab.imshow(x_Ret)
	pylab.show()
	pylab.savefig('converted//' + str(k) + '.png')
	pylab.clf()